providers/certificate.rb

```diff
@@ -0,0 +1,142 @@
+    password = ", \"#{@new_resource.pfx_password}\"" if file.downcase.end_with?('pfx')
+    hash = "(New-Object System.Security.Cryptography.X509Certificates.X509Certificate2 \"#{file}\"#{password}).GetCertHashString()"
+  else
+    # make sure we have no spaces in the hash string
+    hash = "\"#{@new_resource.source.gsub(/\s/, '')}\""
+  end
+  
+  codeScript = <<-EOH
+    #{getAclScript(hash)}
+  EOH
+  onlyifScript = <<-EOH
+    $hash = #{hash}
+    Test-Path "cert:\\#{@location}\\#{@new_resource.store_name}\\$hash"
+  EOH
+  
+  powershell_script @new_resource.name do
+    code codeScript
+    only_if onlyifScript
+  end
+end
+
+action :delete do
+  # We can do everything in a powershell script resource
+  subject = @current_resource.source.sub(/\*/, '`*') # escape any * in the source
+  certCommand = "gci cert:\\#{@location}\\#{@new_resource.store_name} | where { $_.Subject -like '*#{subject}*' -or $_.Thumbprint -eq '#{subject}'}"
+  
+  codeScript = <<-EOH
+    $store = New-Object System.Security.Cryptography.X509Certificates.X509Store "#{@new_resource.store_name}", ([System.Security.Cryptography.X509Certificates.StoreLocation]::#{@location})
+    $store.Open([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadWrite)
+    foreach ($c in #{certCommand})
+    {
+      $store.Remove($c)
+    }
+    $store.Close()
+  EOH
+  onlyifScript = <<-EOH
+    (#{certCommand}).Count -gt 0
+  EOH
+  
+  powershell_script @new_resource.name do
+    code codeScript
+    only_if onlyifScript
+  end
+end
+
+def load_current_resource
+  @current_resource = Chef::Resource::WindowsCertificate.new(@new_resource.name)
+  @current_resource.source(@new_resource.source)
+  @current_resource.pfx_password(@new_resource.pfx_password)
+  @current_resource.private_key_acl(@new_resource.private_key_acl)
+  @current_resource.store_name(@new_resource.store_name)
+  @current_resource.user_store(@new_resource.user_store)
+  @location = @current_resource.user_store ? 'CurrrentUser' : 'LocalMachine'
+end
+
+private
+def getAclScript(hash)
+  if (!@new_resource.private_key_acl.nil? && @new_resource.private_key_acl.length > 0)
+    raise "ACL can only be set on local machine certificates" if @new_resource.user_store
+
+    # this PS came from http://blogs.technet.com/b/operationsguy/archive/2010/11/29/provide-access-to-private-keys-commandline-vs-powershell.aspx
+    setAclScript = <<-EOH
+      $hash = #{hash}
+      $storeCert = gci "cert:\\#{@location}\\#{@new_resource.store_name}\\$hash"
+      if ($storeCert -eq $null) { throw 'no key exists.' } 
+      $keyname = $storeCert.PrivateKey.CspKeyContainerInfo.UniqueKeyContainerName
+      if ($keyname -eq $null) { throw 'no private key exists.' } 
+      $fullpath = "$env:ProgramData\\Microsoft\\Crypto\\RSA\\MachineKeys\\$keyname"
+    EOH
+    @new_resource.private_key_acl.each do | name |
+      setAclScript << "$uname='#{name}'; icacls $fullpath /grant $uname`:RX;"
```

Comment:
Oooh...  I think I understand what you're attempting to do.  You're just granting read access to the underlying stored key and this is just a way to find out the real path.  We have idempotent acl support for windows in the file resource.  See https://docs.chef.io/resource_file.html#windows-file-security

Now that I understand that, there's a slightly better way here.  You still need to run that powershell script to get the full path.  For that you would want to use PowershellOut.  But after you get the fullPath (from stdout of that command), you declare a file resource with the correct path and just pass "rights :read_execute, private_key_acl" to give that array to the resource and let Chef handle it.

If this seems like too big a change, you can leave it as is and I can come back to it during a clean up pass.